org.osgi.service.featurelauncher/src/org/osgi/service/featurelauncher/ArtifactRepositoryFactory.java

```diff
@@ -0,0 +1,64 @@
+/*******************************************************************************
+ * Copyright (c) Contributors to the Eclipse Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *******************************************************************************/
+package org.osgi.service.featurelauncher;
+
+import java.net.URI;
+import java.nio.file.Path;
+import java.util.Map;
+
+import org.osgi.annotation.versioning.ProviderType;
+
+/**
+ * A {@link ArtifactRepositoryFactory} is used to create implementations of
+ * {@link ArtifactRepository} for one of the built in repository types:
+ * <ul>
+ * <li>Local File System</li>
+ * <li>HTTP repository</li>
+ * </ul>
+ */
+@ProviderType
+public interface ArtifactRepositoryFactory {
+
+	/**
+	 * Create an {@link ArtifactRepository} using the local file system
+	 * 
+	 * @param path a path to the root of a Maven Repository Layout containing
+	 *            installable artifacts
+	 * @return an {@link ArtifactRepository} using the local file system
+	 * @throws IllegalArgumentException if the path does not exist, or exists
+	 *             and is not a directory
+	 * @throws NullPointerException if the path is <code>null</code>
+	 */
+	ArtifactRepository localRepository(Path path);
+
+	/**
+	 * Create an {@link ArtifactRepository} using a remote Maven repository
+	 * acessible via HTTP
+	 * 
+	 * @param uri the URI for the repository. The scheme must be
+	 *            <code>http</code> or <code>https</code>
+	 * @param props the configuration properties for the remote repository. See
+	 *            {@link FeatureLauncherConstants} for standard property names
+	 * @return an {@link ArtifactRepository} using the local file system
+	 * @throws IllegalArgumentException if the uri scheme is not
+	 *             <code>http</code> or <code>https</code>
+	 * @throws NullPointerException if the path is <code>null</code>
+	 */
+	ArtifactRepository remoteRepository(URI uri, Map<String,Object> props);
```

Comment:
```suggestion
	ArtifactRepository remoteRepository(URL uri, Map<String,Object> props);
```
If we require it to be http/https as an URL is an URI that has a supported scheme (in this case http)

Reply:
`URI` is usually preferred over `URL` as `URL` has unpleasant behaviours for `equals` and `hashCode`, as well as not handling escaping well.

It's easy to convert between them, so my preference would be to encourage the use of the more modern type (still Java 1.4) with nicer construction and support for relativizing/resolving paths against the URI.

Reply:
How is this important here? The URI must be convertible into an URL as per contract (especially into an http URL) and every URL can be converted into an URI internally (where the reverse is not true) if desired wanted.

That's why I mentioned that one either should use only `createRepository(URI)` with maybe that *at least* http and file scheme *must* be supported (and it doesn't matter if its local and/or remote) or one should be using URL directly (while its unclear to me the benefit of being restrictive on the factory).

Reply:
> How is this important here?

Implementations will need to repeatedly resolve paths against the base that they are given (e.g. to get to the artifact folder) and `URI` provides this natively. It is unlikely that implementations will use `URL.openConnection` or `URL.openStream` as *much* better HTTP clients exist, even in the JDK. 

As another data point, even the standard Java HTTP Client uses URIs not URLs - https://docs.oracle.com/en%2Fjava%2Fjavase%2F11%2Fdocs%2Fapi%2F%2F/java.net.http/java/net/http/HttpRequest.html#newBuilder(java.net.URI)

> while its unclear to me the benefit of being restrictive on the factory

Some of the benefits are that:

- It limits what is mandatory for the implementation to support. Small lightweight implementations may not want to try to implement SSH tunnelling!
- It provides clarity for the user. If I supply a URI of this form then it will work. If not I will get an exception. Optionality is painful because I can never be sure that what I'm doing will work.
- It doesn't prevent an implementation from providing other ArtifactRepository types that users can instantiate or use, it just makes it clear that they're using something implementation specific when they do.

> That's why I mentioned that one either should use only createRepository(URI)

I would be fine to rename the method and also allow `file` URIs. I would expect to keep a version using a `Path` as it is extremely common to use a locally accessible file system as a repository and this method simplifies that use case.